### PR TITLE
Propagate actions from main DOM to Shadow DOM

### LINKED ISSUE #14860: ActionService can't trigger animations defined in shadow documents on `enter`
## What's the issue?

• Document `x` has an `amp-animation`, used by `<amp-position-observer on="enter:recommendedSlideIn.start">`
• `x` is attached to another document `y` using the `attachShadowDoc` method
• When the `enter` action tries to fire, in runs in the context of the parent document `y`. It uses `getElementById` to try and find the animation, and fails, as it's hidden within the shadow document.

## How do we reproduce the issue?

Example on branch: https://github.com/emarchiori/amphtml/tree/shadow-animation-bug-example, in file `examples/land-see/templates/kim-myeong.amp.html`. Scroll down into the second document and the action will try to fire, and fail to find the animation.

## What browsers are affected?

All

## Which AMP version is affected?

1524613093087

Presumably always broken.

### PR DESCRIPTION
**Pull request by nainar**
Currently if we don't find an element to propagate an action to in the main DOM we error. 
However with <amp-next-page> we are now inserting elements/actions into the shadow DOM which can have actions. 

This makes the change to look for elements in the Shadow DOM that may need actions propagated to them. 

This PR is meant to start a discussion, I will work on adding tests once folks think this approach is worth pursuing. 

Fixes #14860.

### CODE REVIEW

File: extensions/amp-position-observer/0.1/amp-position-observer.js [hunk 1]
```diff
@@ -109,7 +109,15 @@ export class AmpVisibilityObserver extends AMP.BaseElement {
    */
   init_() {
     this.parseAttributes_();
-    this.action_ = Services.actionServiceForDoc(this.element);
+    const doc = this.element.ownerDocument;
+    var allNodes = doc.getElementsByTagName('*');
+    var parent = this.element;
+    for (var i = 0; i < allNodes.length; i++) {
+      if(allNodes[i].shadowRoot) {
+        parent = allNodes[i].shadowRoot
```
Comment (jridgewell) [inline extensions/amp-position-observer/0.1/amp-position-observer.js]: I don't understand what this is trying to do.

File: src/service/action-impl.js [hunk 2]
```diff
@@ -468,8 +467,17 @@ export class ActionService {
...
         // Otherwise, targetType should be an element id.
-        const target = (this.globalTargets_[targetType]) ?
+        let target = (this.globalTargets_[targetType]) ?
           this.root_ : this.root_.getElementById(targetType);
+        if (!target) {
+          const doc = this.root_;
+          var allNodes = doc.getElementsByTagName('*');
+          for (var i = 0; i < allNodes.length; i++) {
+            if(allNodes[i].shadowRoot) {
+              target = allNodes[i].shadowRoot.getElementById(targetType);
```
Comment (jridgewell) [inline src/service/action-impl.js]: Ditto.

File: extensions/amp-animation/0.1/web-animations.js [hunk 3]
```diff
@@ -761,18 +761,37 @@ export class MeasureScanner extends Scanner {
           'Both "selector" and "target" are not allowed');
       targets = this.css_.queryElements(spec.selector);
       if (targets.length == 0) {
-        user().warn(TAG, `Target not found: "${spec.selector}"`);
+        const doc = this.css_.rootNode_;
+        var allNodes = doc.getElementsByTagName('*');
+        for (var i = 0; i < allNodes.length; i++) {
+          if(allNodes[i].shadowRoot) {
```
Comment (jridgewell) [inline extensions/amp-animation/0.1/web-animations.js]: Why is an animation crossing a shadow? Shouldn't these be defined inside it?

File: extensions/amp-animation/0.1/web-animations.js [hunk 4]
```diff
@@ -761,18 +761,37 @@ export class MeasureScanner extends Scanner {
           'Both "selector" and "target" are not allowed');
       targets = this.css_.queryElements(spec.selector);
       if (targets.length == 0) {
-        user().warn(TAG, `Target not found: "${spec.selector}"`);
+        const doc = this.css_.rootNode_;
+        var allNodes = doc.getElementsByTagName('*');
```
Comment (jridgewell) [inline extensions/amp-animation/0.1/web-animations.js]: Ooof, this is all going to be really expensive.

File: extensions/amp-position-observer/0.1/amp-position-observer.js (hunk 1 above)
Comment (nainar) [reply extensions/amp-position-observer/0.1/amp-position-observer.js]: Oops this is dead code.

File: src/service/action-impl.js (hunk 2 above)
Comment (nainar) [reply src/service/action-impl.js]: Currently: 
To see what the target for this action is we only check the elements in the document itself. However if the element being targetted is in the shadow document (as is the case with say a position observer in a page inserted via amp-next-page) we never find the target. 

This fixes that by seeing if there are potential targets in the shadow DOM

File: extensions/amp-animation/0.1/web-animations.js (hunk 4 above)
Comment (nainar) [reply extensions/amp-animation/0.1/web-animations.js]: Yes, this is a naive approach. Potentially smarter approaches is holding a reference to all attached Shadow DOMS. Skips having to find them?

File: extensions/amp-animation/0.1/web-animations.js (hunk 3 above)
Comment (nainar) [reply extensions/amp-animation/0.1/web-animations.js]: The animation is defined in the shadow. But the document being checked `this.css_.rootNode_` is the main document which like in `action_impl.js` is unable to find the target node.

File: extensions/amp-animation/0.1/web-animations.js [hunk 5]
```diff
@@ -761,18 +761,37 @@ export class MeasureScanner extends Scanner {
           'Both "selector" and "target" are not allowed');
       targets = this.css_.queryElements(spec.selector);
       if (targets.length == 0) {
-        user().warn(TAG, `Target not found: "${spec.selector}"`);
+        const doc = this.css_.rootNode_;
```
Comment (aghassemi) [inline extensions/amp-animation/0.1/web-animations.js]: `this.css_.queryElements(spec.selector)` should always be correct as it is aware of shadow boundaries, so there might be a bug there or maybe wrong shadowRoot is being passed to the constructor. This should never need to pierce shadow boundaries.

File: src/service/action-impl.js (hunk 2 above)
Comment (aghassemi) [reply src/service/action-impl.js]: same as above, both `position-observer` and `actions` services are doc-level services and in the `amp-next-page` example, `amp-animation`, amp-position-observer` and the actions are all happening in the same document. This should also never pierce shadow boundaries. The root cause is likely something else (wring root being passed somewhere)

**Comment by nainar** [comment]
Given the comments, I'm going to go back and investigate to see that the correct document is being passed from the `position-observer` to the action.

**Comment by nainar** [comment]
Closing as invalid

**Pull request closed without merging**